Document polls (#218)
discord-userdoccers/discord-userdoccers: pages/resources/message.mdx

| components? ^2^         | array[[message component](/interactions/message-components#component-object) object]               | The message's components (e.g. buttons, select menus)                                                                                     |
| sticker_items?          | array[[sticker item](/resources/sticker#sticker-item-object) object]                               | The message's sticker items                                                                                                               |
| stickers?               | array[[sticker](/resources/sticker#sticker-object) object]                                         | Extra rich information for the message's sticker items; only available in some contexts                                                   |

> | poll? ^2^               | [poll](#poll-structure) object                                                                     | A poll!                                                                                                                                   |

Reviewer: ```suggestion
| poll? ^2^               | [poll](#poll-object) object                                                                     | A poll!                                                                                                                                   |
```

| files[n]? ^1^               | file contents                                                                        | Contents of the file being sent (max 10)                                                                                                                                    |
| payload_json? ^1^           | string                                                                               | JSON-encoded body of non-file params                                                                                                                                        |
| attachments? ^1^            | array[partial [attachment](#attachment-object) object]                               | Partial attachment objects with `filename` and `description` (max 10)                                                                                                       |

> | poll?                       | [poll create request](/resources/poll#poll-create-request-object) object             | A poll!                                                                                                                                                                     |

Reviewer: The above alert that mentions which fields are required should also be updated to mention `poll`. This alert is present in a couple other places in the docs as well.

| Field  | Type                                 | Description                   |
| ------ | ------------------------------------ | ----------------------------- |
| embeds | array[[embed](#embed-object) object] | The found embeds for the URLs |

> ### Poll Object

Reviewer: This shouldn't be under the endpoints header, it should be above it, with only the actual endpoints added here

The poll object has a lot of levels and nested structures. It was also designed
to support future extensibility, so some fields may appear to be more complex than
necessary.

###### Poll Structure

| Field             | Type                                                | Description                                      |
| ----------------- | --------------------------------------------------- | ------------------------------------------------ |

> | question ^1^      | [poll media](#poll-media-object-structure) object   | The question of the poll                         |

Reviewer: ```suggestion
| question ^1^      | [poll media](#poll-media-structure) object   | The question of the poll                         |
```

| answers           | array[[poll answer](#poll-answer-structure) object] | The answers available in the poll                |
| expiry ^2^        | ?IS08601 timestamp                                  | When the poll ends                               |
| allow_multiselect | boolean                                             | Whether a user can select multiple answers       |
| layout_type       | integer                                             | The [layout type](#poll-layout-type) of the poll |
| results?          | [poll results](#poll-results-structure) object      | The results of the poll                          |

^1^ Only `text` is supported.

^2^ `expiry` is marked as nullable to support non-expiring polls in the future, but all polls have an expiry currently.

> ###### Example poll

Reviewer: ```suggestion
###### Example Poll
```

This example should be moved to be under all the structures/enums

```json
{
  "question": {
    "text": "Aliens"
  },
  "answers": [
    {
      "answer_id": 1,
      "poll_media": {
        "text": "Alien"
      }
    },
    {
      "answer_id": 2,
      "poll_media": {
        "text": "Alien 2",
        "emoji": {
          "id": null,
          "name": "👽"
        }
      }
    },
    {
      "answer_id": 3,
      "poll_media": {
        "text": "Alien 3",
        "emoji": {
          "id": "1235169043636424757",
          "name": "dolfies"
        }
      }
    }
  ],
  "expiry": "2024-05-02T10:00:02.039342+00:00",
  "allow_multiselect": true,
  "layout_type": 1,
  "results": {
    "answer_counts": [
      {
        "id": 1,
        "count": 1,
        "me_voted": false
      }
    ],
    "is_finalized": false
  }
}
```

###### Poll Create Structure

This is the request object used when creating a poll across the different endpoints.

> It is similar but not exactly identical to the main [poll object](#poll-structure).

Reviewer: ```suggestion
It is similar but not exactly identical to the main [poll](#poll-structure) object.
```

The main difference is that the request has `duration` which eventually becomes `expiry`.

| Field             | Type                                                | Description                                                               |
| ----------------- | --------------------------------------------------- | ------------------------------------------------------------------------- |
| question ^1^      | [poll media](#poll-media-object-structure) object   | The question of the poll                                                  |

> | answers           | array[[poll answer](#poll-answer-structure) object] | Each of the answers available in the poll, up to 10                       |

Reviewer: ```suggestion
| answers           | array[[poll answer](#poll-answer-structure) object] | Each of the answers available in the poll (max 10) |
```

> | duration          | integer                                             | Number of hours the poll should be open for, up to 7 days                 |

Reviewer: ```suggestion
| duration          | integer                                             | Number of hours the poll should be open for (max 7 days)                 |
```

| allow_multiselect | boolean                                             | Whether a user can select multiple answers                                |

> | layout_type?      | integer                                             | The [layout type](#poll-layout-type) of the poll. Defaults to... DEFAULT! |

Reviewer: ```suggestion
| layout_type?      | integer                                             | The [layout type](#poll-layout-type) of the poll (default `DEFAULT`) |
```

^1^ Only `text` is supported.

###### Poll Layout Type

Different layouts for polls will come in the future. For now though, this value will always be `DEFAULT`.

| Value | Name                   | Description                           |
| ----- | ---------------------- | ------------------------------------- |
| 1     | DEFAULT                | The default layout type               |
| ~~2~~ | ~~IMAGE_ONLY_ANSWERS~~ | ~~Poll answers can have only images~~ |

> ### Poll Media Object
> 
> The poll media object is a common object that backs both the question and answers.
> For now, `question` only supports `text`, while answers can have an optional `emoji`.
> 
> ###### Poll Media Object Structure

Reviewer: ```suggestion
###### Poll Media Object Structure

The poll media object is a common object that backs both the question and answers.
For now, `question` only supports `text`, while answers can have an optional `emoji`.
```

| Field      | Type                                           | Description                                                                       |
| ---------- | ---------------------------------------------- | --------------------------------------------------------------------------------- |
| text? ^1^  | string                                         | The text of the field (max 300 characters for question, 55 characters for answer) |
| emoji? ^2^ | partial [emoji](/resources/emoji#emoji-object) | The emoji of the field                                                            |

^1^ `text` should always be non-null for both questions and answers, but please do not depend on that in the future.

^2^ When creating a poll answer with an emoji, one only needs to send either the `id` (custom emoji) or `name` (default emoji) as the only field.

> ### Poll Answer Object
> 
> The `answer_id` is a number that labels each answer.
> As an implementation detail, it currently starts at 1 for the first answer and goes up sequentially.
> We recommend against depending on this sequence.
> 
> Currently, there is a maximum of 10 answers per poll.
> 
> ###### Poll Answer Structure

Reviewer: ```suggestion
###### Poll Answer Structure

The `answer_id` is a number that labels each answer.
As an implementation detail, it currently starts at 1 for the first answer and goes up sequentially.
We recommend against depending on this sequence.

Currently, there is a maximum of 10 answers per poll.
```

| Field         | Type                                              | Description            |
| ------------- | ------------------------------------------------- | ---------------------- |
| answer_id ^1^ | integer                                           | The ID of the answer   |

> | poll_media    | [poll media](#poll-media-object-structure) object | The data of the answer |

Reviewer: ```suggestion
| poll_media    | [poll media](#poll-media-structure) object | The data of the answer |
```

^1^ When sending, this field is optional.

> ### Poll Results Object
> 
> In a nutshell, this contains the number of votes for each answer.
> 
> The `results` field may be not present in certain responses where, as an implementation detail, we do not fetch the poll results in our backend.
> This should be treated as "unknown results", as opposed to "no results". You can keep using the results if you have previously received them through other means.
> 
> Also due to the intricacies of counting at scale, while a poll is in progress the results may not be perfectly accurate.
> They usually are accurate, and shouldn't deviate significantly—it's just difficult to make guarantees.
> 
> To compensate for this, after a poll is finished there is a background job which performs a final, accurate tally of votes.
> This tally concludes once `is_finalized` is `true`. Polls that have ended will also always contain results.
> 
> If `answer_counts` does not contain an entry for a particular answer, then there are no votes for that answer.
> 
> ###### Poll Results Structure

Reviewer: ```suggestion
###### Poll Results Structure

In a nutshell, this contains the number of votes for each answer.

The `results` field may be not present in certain responses where, as an implementation detail, we do not fetch the poll results in our backend.
This should be treated as "unknown results", as opposed to "no results". You can keep using the results if you have previously received them through other means.

Also due to the intricacies of counting at scale, while a poll is in progress the results may not be perfectly accurate.
They usually are accurate, and shouldn't deviate significantly—it's just difficult to make guarantees.

To compensate for this, after a poll is finished there is a background job which performs a final, accurate tally of votes.
This tally concludes once `is_finalized` is `true`. Polls that have ended will also always contain results.

If `answer_counts` does not contain an entry for a particular answer, then there are no votes for that answer.
```

| count    | integer | The number of votes for this answer            |
| me_voted | boolean | Whether the current user voted for this answer |

<RouteHeader method="GET" url="/channels/{channel.id}/polls/{message.id}/answers/{answer.id}">
  Get Answer Voters
</RouteHeader>

Get a list of users that voted for this specific answer.

###### Query String Params

| Field  | Type      | Description                                       |
| ------ | --------- | ------------------------------------------------- |
| after? | snowflake | Get users after this user ID                      |
| limit? | integer   | Max number of users to return (1-100, default 25) |

###### Response Body

| Field | Type                                                      | Description                     |
| ----- | --------------------------------------------------------- | ------------------------------- |
| users | array[partial [user](/resources/user#user-object) object] | Users who voted for this answer |

<RouteHeader method="POST" url="/channels/{channel.id}/polls/{message.id}/expire">
  End Poll
</RouteHeader>

Immediately ends the poll. You cannot end polls from other users.

> Returns a [message](/resources/message#message-object) object. Fires a [Message Update](/topics/gateway-events#message-update) Gateway event.

Reviewer: ```suggestion
Returns a [message](#message-object) object. Fires a [Message Update](/topics/gateway-events#message-update) Gateway event.
```

<RouteHeader method="PUT" url="/channels/{channel.id}/polls/{message.id}/answers/@me">

>   Submit Poll Vote

Reviewer: ```suggestion
  Create Poll Vote
```

</RouteHeader>

> Submit [poll](#poll-structure) votes. Returns a 204 empty response on success. Fires multiple [Message Poll Vote Add](/topics/gateway-events#message-poll-vote-add)/[Message Poll Vote Remove](/topics/gateway-events#message-poll-vote-remove) Gateway events.

Reviewer: ```suggestion
Submits a poll vote for the current user. Returns a 204 empty response on success. Fires multiple [Message Poll Vote Add](/topics/gateway-events#message-poll-vote-add) and optionally [Message Poll Vote Remove](/topics/gateway-events#message-poll-vote-remove) Gateway events.
```

###### JSON Params

| Field      | Type          | Description                            |
| ---------- | ------------- | -------------------------------------- |

> | answer_ids | array[string] | Selected answers, empty to clear votes |

Reviewer: ```suggestion
| answer_ids | array[integer] | Selected answers, empty to clear votes |
```

These are actually integers, the client just casts them into strings because reasons

> 
> ###### Example Request Body
> 
> ```json
> {
>   "answer_ids": ["1"]
> }
> ```

Reviewer: ```suggestion
```
